### PR TITLE
Better Unicode input support on Windows systems.

### PR DESCRIPTION
**Pull request by samhocevar**
- support surrogate pairs in WM_CHAR messages
- support characters from supplementary planes in WM_UNICHAR messages
- support WM_UNICHAR messages in non-Unicode mode
- fix some code duplication

Also, do not return FALSE upon receiving a WM_UNICHAR message with
a UNICODE_NOCHAR parameter, since WM_UNICHAR can actually be handled
just fine (at least with the exact same level of support as WM_CHAR).

### CODE REVIEW

**Comment by samhocevar** [comment]
I got this fix [pushed to upstream scintilla](https://sourceforge.net/p/scintilla/code/ci/ce4680b7180549ac118893aff9034214c412f193/) so I guess this pull request is no longer relevant.

**Comment by samhocevar** [comment]
Hi, it’s been two years and #1126 is still open. Maybe you’d be willing to apply this PR instead in the meantime? Emoji are all over the place nowadays and Notepad++ still does not allow to input them through keyboard input messages. 😢😢😢

**Comment by Fuzzyma** [comment]
Jep - I recently came acroos this issue. Instead of eh imoji a white recatngle appears. But even worse: After saving and loading, the rectagle now was cryptic chars with an invisible escape sequence in front.
That was no fun to fix! Why is notepad++ not falling back on a font which supports emojis?

File: scintilla/win32/ScintillaWin.cxx [hunk 1]
```diff
@@ -1398,40 +1428,32 @@ sptr_t ScintillaWin::WndProc(unsigned int iMessage, uptr_t wParam, sptr_t lParam
...
+					wclen = 2;
 				}
+				AddCharUTF16(wcs, wclen);
 			}
 			return 0;
 
 		case WM_UNICHAR:
 			if (wParam == UNICODE_NOCHAR) {
-				return IsUnicodeMode() ? 1 : 0;
+				return TRUE;
```
Comment (MAPJe71) [inline scintilla/win32/ScintillaWin.cxx]: Please keep return type numeric or at least use approppriate C++ typecast to return type `sptr_t`.

File: scintilla/win32/ScintillaWin.cxx [hunk 2]
```diff
@@ -1398,40 +1428,32 @@ sptr_t ScintillaWin::WndProc(unsigned int iMessage, uptr_t wParam, sptr_t lParam
...
-					UTF8FromUTF16(wcs, 1, utfval, len);
-					AddCharUTF(utfval, len);
-					return 1;
-				} else {
-					return 0;
-				}
+				wchar_t wcs[3] = {0};
+				unsigned int wclen = UTF16FromUTF32Character(wParam, wcs);
+				AddCharUTF16(wcs, wclen);
+				return FALSE;
```
Comment (MAPJe71) [inline scintilla/win32/ScintillaWin.cxx]: Likewise.

File: scintilla/win32/ScintillaWin.cxx (hunk 1 above)
Comment (samhocevar) [reply scintilla/win32/ScintillaWin.cxx]: @MAPJe71 fixed! Thanks for looking at the patch.

**Comment by donho** [comment]
This PR should be submitted in Scintilla project, but not here.

**Comment by samhocevar** [comment]
@donho As mentioned above, this PR *has* been submitted (and accepted) in Scintilla **two years ago**, but your version of Scintilla is lagging three years behind (see #1126), hence my question above:

  >Maybe you’d be willing to apply this PR instead in the meantime?

Also, I find it a bit rude to close a PR after I’ve been asked to perform extra work on it…

Please let me know how I can help Notepad++ get proper Unicode input support, what kind of work is necessary? I’ve been trying for years now.

**Pull request closed without merging**